[IBM WebSphere Application Server] Add WebSphere Application Server Integration Package with JDBC data stream

kush-elastic opened the pull request:
- Enhancement

## What does this PR do?

- Generated the skeleton of WebSphere Application Server integration package.
- Added 1 data stream (JDBC).
- Added data collection logic for the data streams.
- Added the ingest pipeline for the data streams.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yaml files.
- Added dashboards and visualizations.
- Added pipeline test cases for the data stream.
- Added system test cases for the data stream.

## Checklist

- [X] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [X] I have verified that all data streams collect metrics or logs.
- [X] I have added an entry to my package's changelog.yml file.
- [X] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/websphere_application_server directory.
- Run the following command to run tests.
`elastic-package test`

## Related issues

- Relates #2847
- Relates #3085
- Relates #3138
- Relates #3199

## Screenshots
![websphere_application_server-jdbc-dashboard-screenshot](https://user-images.githubusercontent.com/89848966/163380731-a0fe3394-80b5-42ac-90bd-b36522a44567.png)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2922/detail/PR-2922/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2922/detail/PR-2922/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2922/detail/PR-2922/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2922/detail/PR-2922/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_2922.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-03-30T11:55:02.278Z&rangeTo=2022-03-30T12:15:02.278Z&transactionName=BUILD+Ingest-manager/integrations/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=7135b6c127b6bf4a57cc8fcdcdaad216&transactionId=7e070d2380f373bc)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-03-30T12:05:02.278+0000


* Duration: 4 min 41 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-2922/detail/PR-2922/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Checks and builds Go sources`
<ul>
<li>Took 0 min 14 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-2922/runs/1/steps/140/log/?start=0">here</a></li>
<li>Description: <code>mage -debug check</code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 0 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-2922/runs/1/steps/157/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

kush-elastic commented:
/test

yug-elastic commented on packages/websphere_application_server/docs/README.md:
```diff
@@ -0,0 +1,152 @@
+# WebSphere Application Server
+
+This Elastic integration is used to collect [IBM WebSphere Application Server](https://www.ibm.com/cloud/websphere-application-server) metrics as follows:
```
nit: 
```
This Elastic integration is used to collect the following metrics from [IBM WebSphere Application Server](https://www.ibm.com/cloud/websphere-application-server):
```

yug-elastic commented on packages/websphere_application_server/manifest.yml:
```diff
@@ -0,0 +1,92 @@
+title: WebSphere Application Server
+version: 0.1.0
+license: basic
+description: This Elastic integration collects metrics from WebSphere Application Server.
```
Please adjust the description as per https://github.com/elastic/integrations/pull/2614#discussion_r860162158

yug-elastic commented on packages/websphere_application_server/manifest.yml:
```diff
@@ -0,0 +1,92 @@
+            title: Host
+            required: true
+            show_user: true
+            description: WebSphere Application Server Host. (http://localhost:9080 or localhost:9080).
```
nit: `WebSphere Application Server Host (http://localhost:9080 or localhost:9080).`

yug-elastic reviewed (changes requested):
Requested some changes

kush-elastic commented on packages/websphere_application_server/data_stream/jdbc/_dev/test/pipeline/test-jdbc-metrics.json-expected.json:
```diff
@@ -0,0 +1,127 @@
+                "address": "10.0.14.233:9080"
+            },
+            "tags": [
+                "prometheus"
```
We're giving user input [tags](https://github.com/kush-elastic/integrations/blob/package_websphere_application_server_jdbc/packages/websphere_application_server/data_stream/jdbc/manifest.yml#L14) where default value contains this tag if user want to keep it.

yug-elastic replied:
Got it, thanks for clarifying!

yug-elastic reviewed (approved):
Thanks for making the adjustments, @kush-elastic. LGTM!
Please make the changes suggested in this PR (if applicable) to the other IBM WAS PRs as well.

kush-elastic commented:
/test

kush-elastic commented on packages/websphere_application_server/data_stream/jdbc/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,161 @@
+  - rename:
+      field: prometheus.metrics.was_connectionpool_jdbcOperation_time_seconds_total
+      target_field: websphere_application_server.jdbc.connection.total.operations_seconds
+      ignore_missing: true
```
Elastic has a standard field naming convention: https://www.elastic.co/guide/en/beats/devguide/current/event-conventions.html and we've been following the same standards throughout all the integration. All the fields that we are extracting in this integration also follows the same naming standards. Please refer to the shared doc and let me know if you still think we should change it.
CC: @akshay-saraswat

ishleenk17 replied:
Thanks for the clarification.

kush-elastic commented on packages/websphere_application_server/data_stream/jdbc/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,161 @@
+  - rename:
+      field: prometheus.metrics.was_connectionpool_jdbcOperation_time_seconds_total
+      target_field: websphere_application_server.jdbc.connection.total.operations_seconds
+      ignore_missing: true
```
Please refer to the reply: https://github.com/elastic/integrations/pull/2922#discussion_r875824695

kush-elastic commented on packages/websphere_application_server/data_stream/jdbc/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,161 @@
+      ignore_missing: true
+  - script:
+      description: Drops null/empty values recursively
+      lang: painless
```
I guess you should [refer this](https://github.com/elastic/integrations/pull/3066#discussion_r866704272) for drop processor discussion.

kush-elastic commented on packages/websphere_application_server/_dev/deploy/docker/docker-compose.yml:
```diff
@@ -0,0 +1,15 @@
+version: '3.4'
+services:
+  websphere_application_server:
+    image: ibmcom/websphere-traditional:9.0.5.10
```
Sure, I have update new image. Please check it out.

ishleenk17 commented on packages/websphere_application_server/data_stream/jdbc/fields/base-fields.yml:
```diff
@@ -0,0 +1,12 @@
+  description: Data stream namespace.
+- name: '@timestamp'
+  type: date
+  description: Event timestamp.
```
It is always a good practice to have the event.* fields have their values populated in base-fields.yml.
Please refer to this as example:
https://github.com/elastic/integrations/blob/main/packages/prometheus/data_stream/collector/fields/base-fields.yml

kush-elastic replied:
We have followed [this](https://www.elastic.co/guide/en/integrations-developer/current/add-a-mapping.html#:~:text=base%2Dfields.yml%20never%20changes%20and%20is%20required%20for%20all%20integrations) from integration developer guide.
if you still think we should change this please let us know.

ishleenk17 replied:
This would mean that the existing fields in it will never change.
But the newer fields can be added. That's what is followed through all the integrations in Elastic.

kush-elastic replied:
Okay, Thanks. will update it soon.

kush-elastic commented:
Hey @ishleenk17, thanks for the review! We have addressed all the review comments and made the required changes and re-requested review. Could you please take a look at the PR again and let us know if you think we should change anything else before we can have an approval from you over the PR?

ishleenk17 commented:
Looks good. Approved!
